lib/streamaggr/total.go

```diff
@@ -2,185 +2,100 @@ package streamaggr
 
 import (
 	"math"
-	"sync"
-	"time"
 
 	"github.com/VictoriaMetrics/VictoriaMetrics/lib/bytesutil"
-	"github.com/VictoriaMetrics/VictoriaMetrics/lib/fasttime"
 )
 
-// totalAggrState calculates output=total, total_prometheus, increase and increase_prometheus.
-type totalAggrState struct {
-	m sync.Map
-
-	// Whether to reset the output value on every flushState call.
-	resetTotalOnFlush bool
-
-	// Whether to take into account the first sample in new time series when calculating the output value.
-	keepFirstSample bool
-
-	// Time series state is dropped if no new samples are received during stalenessSecs.
-	//
-	// Aslo, the first sample per each new series is ignored during stalenessSecs even if keepFirstSample is set.
-	// see ignoreFirstSampleDeadline for more details.
-	stalenessSecs uint64
-
-	// The first sample per each new series is ignored until this unix timestamp deadline in seconds even if keepFirstSample is set.
-	// This allows avoiding an initial spike of the output values at startup when new time series
-	// cannot be distinguished from already existing series. This is tracked with ignoreFirstSampleDeadline.
-	ignoreFirstSampleDeadline uint64
-}
-
-type totalStateValue struct {
-	mu             sync.Mutex
-	lastValues     map[string]totalLastValueState
-	total          float64
-	deleteDeadline uint64
-	deleted        bool
+func totalInitFn(resetTotalOnFlush, keepFirstSample bool) aggrValuesInitFn {
+	return func(values []aggrValue) []aggrValue {
+		shared := &totalAggrValueShared{
+			lastValues: make(map[string]totalLastValue),
+		}
+		for i := range values {
+			values[i] = &totalAggrValue{
+				keepFirstSample:   keepFirstSample,
+				resetTotalOnFlush: resetTotalOnFlush,
+				shared:            shared,
+			}
+		}
+		return values
+	}
 }
 
-type totalLastValueState struct {
+type totalLastValue struct {
 	value          float64
 	timestamp      int64
-	deleteDeadline uint64
+	deleteDeadline int64
 }
 
-func newTotalAggrState(stalenessInterval time.Duration, resetTotalOnFlush, keepFirstSample bool) *totalAggrState {
-	stalenessSecs := roundDurationToSecs(stalenessInterval)
-	ignoreFirstSampleDeadline := fasttime.UnixTimestamp() + stalenessSecs
-
-	return &totalAggrState{
-		resetTotalOnFlush:         resetTotalOnFlush,
-		keepFirstSample:           keepFirstSample,
-		stalenessSecs:             stalenessSecs,
-		ignoreFirstSampleDeadline: ignoreFirstSampleDeadline,
-	}
+type totalAggrValueShared struct {
+	lastValues map[string]totalLastValue
+	total      float64
 }
 
-func (as *totalAggrState) pushSamples(samples []pushSample) {
-	currentTime := fasttime.UnixTimestamp()
-	deleteDeadline := currentTime + as.stalenessSecs
-	keepFirstSample := as.keepFirstSample && currentTime > as.ignoreFirstSampleDeadline
```

Comment:
Will there be any alternative of `ignoreFirstSampleDeadline` with aggregation windows? Asking because it seems now they are missing with this PR.

Reply:
we're considering removing it in another PR

Reply:
https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7116 ?
I'm worried that there is removing of quite useful feature.